### PR TITLE
Expose coordinator's internal functionality to manage cluster

### PR DESCRIPTION
**Pull request by antoniofilipovic**
### Description

This diff adds the option for users to run actions like DEMOTE INSTANCE and FORCE RESET CLUSTER STATE, which are run internally but now open for users to reset cluster state, and choose their instance to be MAIN.

TODO:
- [ ] learn how to trim commits in merge to master, ping @as51340 
- [x] refactor force reset cluster 

[master < Task] PR
- [x] Provide the full content or a guide for the final git message
    - Expose some internal functionality to manage cluster

### CI Testing Labels
Please select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_

### Documentation checklist
- [x] Add the documentation label tag
- [x] Add the bug / feature label tag
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [x] Write a release note, including added/changed clauses
    - You can now run demote instance to replica and force reset cluster
- [ ] Link the documentation PR here
    - **[Documentation PR link]**
- [x] Tag someone from docs team in the comments

### CODE REVIEW

**Comment by antoniofilipovic** [comment]
@kgolubic will add docs PR soon

File: src/coordination/coordinator_state.cpp [hunk 1]
```diff
@@ -11,19 +11,19 @@
 
 #ifdef MG_ENTERPRISE
 
-#include "coordination/coordinator_state.hpp"
+#include <optional>
+#include <string_view>
+#include <variant>
+
```
Comment (as51340) [inline src/coordination/coordinator_state.cpp:17]: I don't care about this but I think the usual order is:
1. header file
2. other our files
3. libraries

If you can verify with the other files, it's good if we are consistent

File: tests/e2e/high_availability/distributed_coords.py [hunk 2]
```diff
@@ -2504,5 +2504,275 @@ def get_vertex_count():
...
+    mg_sleep_and_assert(leader_data, show_instances_coord3)
+    mg_sleep_and_assert(follower_data, show_instances_coord1)
+    mg_sleep_and_assert(follower_data, show_instances_coord2)
+
+    with pytest.raises(Exception) as e:
+        execute_and_fetch_all(instance_3_cursor, "SHOW REPLICAS;")
+    assert str(e.value) == "Replica can't show registered replicas (it shouldn't have any)!"
+
+    # We are explicitly waiting here if failover will happen - we expect it shouldn't happen
+    time.sleep(FAILOVER_PERIOD + 1)
```
Comment (as51340) [inline tests/e2e/high_availability/distributed_coords.py:2635]: Can we write test without this -> we know it will become flaky without it

File: src/query/interpreter.cpp [hunk 3]
```diff
@@ -432,6 +434,93 @@ class CoordQueryHandler final : public query::CoordinatorQueryHandler {
...
+  void ForceResetClusterState() override {
+    auto status = coordinator_handler_.ForceResetClusterState();
+    switch (status) {
+      using enum memgraph::coordination::ForceResetClusterStateStatus;
+      case NOT_COORDINATOR:
+        throw QueryRuntimeException("Force reset cluster state query can only be run on a coordinator!");
+      case SHUTTING_DOWN:
+        throw QueryRuntimeException("Force reset cluster aborted as coordinator is shutting down!");
+      case ACTION_FAILED:
+        throw QueryRuntimeException("Check logs for more details, one action didn't complete successfully!");
```
Comment (as51340) [inline src/query/interpreter.cpp:488]: Which action? Improve message

File: src/memgraph.cpp [hunk 4]
```diff
@@ -669,6 +669,9 @@ int main(int argc, char **argv) {
     websocket_server.Shutdown();
 #ifdef MG_ENTERPRISE
     metrics_server.Shutdown();
+    if (coordinator_state.has_value() && coordinator_state->IsCoordinator()) {
+      coordinator_state->ShutDownCoordinator();
+    }
```
Comment (as51340) [inline src/memgraph.cpp:674]: Why are we doing this?

File: src/coordination/include/coordination/coordinator_instance.hpp [hunk 5]
```diff
@@ -130,25 +138,24 @@ class CoordinatorInstance {
...
   mutable utils::ResourceLock coord_instance_lock_{};
 
-  // Thread pool needs to be constructed before raft state as raft state can call thread pool
-  utils::ThreadPool thread_pool_{1};
-
-  // This needs to be constructed last because raft state may call
-  // setting up leader instance
   std::unique_ptr<RaftState> raft_state_;
+
+  // Thread pool must be destructed first, because there is option we are doing force reset in thread
```
Comment (as51340) [inline src/coordination/include/coordination/coordinator_instance.hpp:156]:
```suggestion
  // Thread pool must be destructed first, because there is a possibility we are doing force reset in thread
```

File: src/coordination/coordinator_instance.cpp [hunk 6]
```diff
@@ -66,11 +67,28 @@ CoordinatorInstance::CoordinatorInstance(CoordinatorInstanceInitConfig const &co
   raft_state_->InitRaftServer();
 }
 
+CoordinatorInstance::~CoordinatorInstance() {
+  // Order is important:
+  // 1. Thread pool might be running with force reset, or become follower, so we should shut that down
+  // 2. Await shutdown of coordinator thread pool so that coordinator can't become follower or do force reset
+  // 3. Frequent checks running, we need to stop them before raft_state_ is destroyed
+  ShuttingDown();
+  thread_pool_.Shutdown();
```
Comment (as51340) [inline src/coordination/coordinator_instance.cpp:76]: For consistency rename `Shutdown` to `ShutDown`

File: src/coordination/include/coordination/coordinator_instance.hpp [hunk 7]
```diff
@@ -74,6 +74,14 @@ class CoordinatorInstance {
 
   auto GetLeaderCoordinatorData() const -> std::optional<CoordinatorToCoordinatorConfig>;
 
+  auto DemoteInstanceToReplica(std::string_view instance_name) -> DemoteInstanceCoordinatorStatus;
+
+  auto TryForceResetClusterState() -> ForceResetClusterStateStatus;
```
Comment (as51340) [inline src/coordination/include/coordination/coordinator_instance.hpp:79]: Where is this method used? Can we have `ForceResetClusterState` and its private method

File: src/coordination/coordinator_instance.cpp [hunk 8]
```diff
@@ -395,14 +415,49 @@ void CoordinatorInstance::ForceResetCluster() {
...
+      case ForceResetClusterStateStatus::SHUTTING_DOWN:
+        spdlog::trace("Stopping force as coordinator is shutting down.");
+        return result;
+      case ForceResetClusterStateStatus::RPC_FAILED:
+      case ForceResetClusterStateStatus::ACTION_FAILED:
+      case ForceResetClusterStateStatus::RAFT_LOG_ERROR:
+      case ForceResetClusterStateStatus::NO_NEW_MAIN:
+      case ForceResetClusterStateStatus::FAILED_TO_OPEN_LOCK:
+      case ForceResetClusterStateStatus::FAILED_TO_CLOSE_LOCK:
+      case ForceResetClusterStateStatus::NOT_COORDINATOR:  // shouldn't happen
```
Comment (as51340) [inline src/coordination/coordinator_instance.cpp:451]: Can we annotate with [[fallthrough]]?

File: src/coordination/coordinator_instance.cpp [hunk 9]
```diff
@@ -670,6 +725,61 @@ auto CoordinatorInstance::RegisterReplicationInstance(CoordinatorToReplicaConfig
   return RegisterInstanceCoordinatorStatus::SUCCESS;
 }
 
+auto CoordinatorInstance::DemoteInstanceToReplica(std::string_view instance_name) -> DemoteInstanceCoordinatorStatus {
+  auto lock = std::lock_guard{coord_instance_lock_};
+
+  if (!is_leader_ready_) {
```
Comment (as51340) [inline src/coordination/coordinator_instance.cpp:731]: When the leader isn't ready?

File: src/coordination/include/coordination/coordinator_instance.hpp [hunk 10]
```diff
@@ -130,25 +138,24 @@ class CoordinatorInstance {
...
-  void ForceResetCluster();
+  auto ForceResetCluster_() -> ForceResetClusterStateStatus;
 
   auto GetBecomeLeaderCallback() -> std::function<void()>;
   auto GetBecomeFollowerCallback() -> std::function<void()>;
 
   HealthCheckClientCallback client_succ_cb_, client_fail_cb_;
   std::atomic<bool> is_leader_ready_{false};
-
+  std::atomic<bool> is_shutting_down_{false};
```
Comment (as51340) [inline src/coordination/include/coordination/coordinator_instance.hpp:148]: Can you write here when is_leader_ready_ and when not?

File: src/coordination/include/coordination/coordinator_instance.hpp [hunk 11]
```diff
@@ -130,25 +138,24 @@ class CoordinatorInstance {
...
   // TODO(antoniofilipovic) do we still rely on pointer stability
   std::list<ReplicationInstanceConnector> repl_instances_;
   mutable utils::ResourceLock coord_instance_lock_{};
 
-  // Thread pool needs to be constructed before raft state as raft state can call thread pool
-  utils::ThreadPool thread_pool_{1};
-
-  // This needs to be constructed last because raft state may call
-  // setting up leader instance
   std::unique_ptr<RaftState> raft_state_;
```
Comment (as51340) [inline src/coordination/include/coordination/coordinator_instance.hpp:154]: Does this condition still apply or? When does thread pool need to be constructed?

File: src/coordination/coordinator_instance.cpp (hunk 6 above)
Comment (as51340) [inline src/coordination/coordinator_instance.cpp:76]: Why this explicit handling of shutdown? Why is this necessary, for what?
Can you write also in code comment about that. CoordinatorInstance has become complex so maybe it wouldn't hurt

**Review by as51340 — CHANGES_REQUESTED** [review]
good work, I don't understand some things

**Comment by as51340** [comment]
What is force reset action doing?

**Comment by as51340** [comment]
Can we name better this PR? "some internal functionality" to something more meaningful

File: tests/e2e/high_availability/distributed_coords.py (hunk 2 above)
Comment (antoniofilipovic) [reply tests/e2e/high_availability/distributed_coords.py:2635]: I can change it to have mg_sleep_and_assert_for_a_while so that we wait for 3 seconds and check that state is consistent.

File: src/query/interpreter.cpp (hunk 3 above)
Comment (antoniofilipovic) [reply src/query/interpreter.cpp:488]: There are two different actions that can fail, for example sending SwapUUID, DemoteToReplica. I wanted to put them under same bin, but I will see if I can split it and introduce different error message for those parts

File: src/memgraph.cpp (hunk 4 above)
Comment (antoniofilipovic) [reply src/memgraph.cpp:674]: Coordinator might be in infinite loop doing ForceReset and it might be stuck and we might not shut down it at any point soon. So we are doing shutdown this way

File: src/coordination/include/coordination/coordinator_instance.hpp (hunk 7 above)
Comment (antoniofilipovic) [reply src/coordination/include/coordination/coordinator_instance.hpp:79]: This is user action for force reset

File: src/coordination/coordinator_instance.cpp (hunk 9 above)
Comment (antoniofilipovic) [reply src/coordination/coordinator_instance.cpp:731]: If it is just becoming leader, RAFT server elected it

File: src/memgraph.cpp (hunk 4 above)
Comment (antoniofilipovic) [reply src/memgraph.cpp:674]: I will write on force reset why are we doing this

File: src/coordination/include/coordination/coordinator_instance.hpp (hunk 7 above)
Comment (antoniofilipovic) [reply src/coordination/include/coordination/coordinator_instance.hpp:79]: Write that down

File: src/coordination/coordinator_instance.cpp (hunk 9 above)
Comment (antoniofilipovic) [reply src/coordination/coordinator_instance.cpp:731]: write that down

File: src/coordination/coordinator_state.cpp (hunk 1 above)
Comment (antoniofilipovic) [reply src/coordination/coordinator_state.cpp:17]: I think we have incosistentices in our codebase: 
In memgraph.cpp, order of includes is:
c++ libraries
header files
3rd party libraries

Interpreter.cpp is chaos a bit. storage.cpp also. I will change as you mentioned here, I think we can raise this on our weekly meeting

File: tests/e2e/high_availability/distributed_coords.py (hunk 2 above)
Comment (antoniofilipovic) [reply tests/e2e/high_availability/distributed_coords.py:2635]: Done

File: src/query/interpreter.cpp (hunk 3 above)
Comment (antoniofilipovic) [reply src/query/interpreter.cpp:488]: resolved on call not to include message

File: src/coordination/include/coordination/coordinator_instance.hpp (hunk 11 above)
Comment (antoniofilipovic) [reply src/coordination/include/coordination/coordinator_instance.hpp:154]: Explained on call

File: src/coordination/coordinator_instance.cpp (hunk 6 above)
Comment (antoniofilipovic) [reply src/coordination/coordinator_instance.cpp:76]: done

File: tests/e2e/mg_utils.py [hunk 12]
```diff
@@ -17,6 +17,19 @@ def mg_sleep_and_assert(expected_value, function_to_retrieve_data, max_duration=
     return result
 
 
+def mg_sleep_and_assert_for_a_while(
```
Comment (as51340) [inline tests/e2e/mg_utils.py:20]: Rename this to `mg_assert_after`

Comment (antoniofilipovic) [reply tests/e2e/mg_utils.py:20]: Changed to `mg_assert_until` as we need to ping for some time
